### PR TITLE
[`LoRA`] Fix lora + PEFT issues

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Currently the CI of PEFT is failing on main, this PR is an attempt to fix the failing CI on PEFT while preserving the previous behaviour on diffusers side. 

As of now `LoRACompatibleLinear` and `LoRACompatibleConv` respectively inherit from `nn.Linear` and `nn.Conv2d` layers and in the modeling code, `forward` method is called with optional arguments that are explicitly passed. This causes some issues on PEFT as we make an assumption that for linear and conv2d modules the forward method signature is not changed. 

Although it is possible to make PEFT layers more flexible (e.g. https://github.com/huggingface/peft/pull/933) to accept more arguments at forward, it still breaks some scope of usage of PEFT + diffusers. For example when calling `merge_and_unload` on a unet module, all `LoRACompatibleConv` and `LoRACompatibleLinear` layers are respectively replaced with `nn.Linear` and `nn.Conv2d` (since they inherit from these classes), but the merged model will break at inference since `scale` is explicitly passed to the forward method. 

In this (*hacky*) PR I propose to revisit the approach currently taken by making changes mathematically. 

As pointed out by patrick offline, the scale parameter that `LoRACompatiblexxx` takes at forward is different from the scale parameter of LoRA. However, mathematically we should be able to retrieve the same results  if we simply multiply the `self.lora_layer.network_alpha` by that scale, and make sure to remove its effect by diving `self.lora_layer.network_alpha` with that scale after the forward pass. 

If no network alpha is set, we can simply set network alpha to `scale ** 2` and the rank parameter to `scale` to emulate the operation we want to perform for `LoRACompatiblexxx`

First multiplication here: https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/lora.py#L66 
Second multiplication here: https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/lora.py#L227

This PR proposes a hacky solution, however it is the only I have found so far to resolve the issue on PEFT side while keeping things working as before on diffusers

### CODE REVIEW

File: src/diffusers/models/attention.py [hunk 1]
```diff
@@ -303,8 +303,12 @@ def __init__(
 
     def forward(self, hidden_states, scale: float = 1.0):
         for module in self.net:
-            if isinstance(module, (LoRACompatibleLinear, GEGLU)):
+            if isinstance(module, (GEGLU)):
```
Comment (sayakpaul) [inline src/diffusers/models/attention.py:306]: nit: no need to use a tuple for a single type check.

File: src/diffusers/models/attention.py [hunk 2]
```diff
@@ -303,8 +303,12 @@ def __init__(
...
                 hidden_states = module(hidden_states, scale)
+            elif isinstance(module, (LoRACompatibleLinear)) and scale != 1.0:
+                module.add_scale(scale)
+                hidden_states = module(hidden_states)
+                module.remove_scale(scale)
```
Comment (sayakpaul) [inline src/diffusers/models/attention.py:311]: An explanatory comment will be useful.

File: src/diffusers/models/attention_processor.py [hunk 3]
```diff
@@ -542,6 +542,11 @@ def __call__(
     ):
         residual = hidden_states
 
+        if isinstance(scale, float) and scale != 1.0:
+            for module in attn.modules():
+                if isinstance(module, (LoRACompatibleLinear, LoRACompatibleConv)):
+                    module.add_scale(scale)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:548]: Shouldn't we also have a corresponding `remove_scale()` here?

File: src/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -590,6 +595,11 @@ def __call__(
 
         hidden_states = hidden_states / attn.rescale_output_factor
 
+        if isinstance(scale, float) and scale != 1.0:
+            for module in attn.modules():
+                if isinstance(module, (LoRACompatibleLinear, LoRACompatibleConv)):
+                    module.remove_scale(scale)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:601]: Apologies if I missed it out but I didn't see a corresponding `add_scale()`.

**Review by sayakpaul — COMMENTED** [review]
Thanks so much for thinking this through! Not an easy one but I think you tried your best to find the right balance between being mathematically sound and API consistency. 

My thoughts:

* Would be nice to have a simple equation / worked out example that validates your formulation of how you're using `scale` as a part of the PR description. 
* The code currently reads slightly convoluted, but I guess that's the compromise we'll have to settle with. 
* Let's ensure we have elaborate comments to explain why we're doing `add_scale()` and `remove_scale()`.

File: src/diffusers/models/lora.py [hunk 5]
```diff
@@ -156,15 +155,28 @@ def _unfuse_lora(self):
         self.w_up = None
         self.w_down = None
 
-    def forward(self, hidden_states, scale: float = 1.0):
+    def add_scale(self, scale: float = 1.0):
```
Comment (patrickvonplaten) [inline src/diffusers/models/lora.py:158]: This would break the `fuse` and `unfuse` operation. We cannot change `network_alpha` on the fly

**Comment by patrickvonplaten** [comment]
This is a bit too much PEFT-related changes directly in `diffusers`. 
- It breaks `fuse` and `unfuse`
- It makes an already complex code even more complex

I don't really understand what PEFT is doing there under the hood, but if such changes are required in `diffusers` than it's really too much black magic stuff.

Can we maybe for now just pin `diffusers` to the previous version in PEFT and fully focus on PEFT<>Diffusers integration?

**Comment by younesbelkada** [comment]
Yes this approach is very hacky as I said :/ 
OK for me to fully focus on PEFT + diffusers and pin diffusers version in PEFT @pacman100 @BenjaminBossan

**Pull request closed without merging**